File: JavaScript/面试题 17.17. 多次搜索.js
/**
 * @param {string} big
 * @param {string[]} smalls
 * @return {number[][]}
 */
 var multiSearch = function(big, smalls) {
    var dict=new Map();
    for(var i=0;i<big.length;i++){
        var temp="";
        for(var j=i;j<big.length;j++){
            temp+=big[j];
            //console.log(temp);
            if(!dict.has(temp)){
                //console.log(temp);
                dict.set(temp,[i]);
            }else{
                var l=dict.get(temp);
                l.push(i);
                dict.set(temp,l);
            }
        }
        // console.log(dict);
    }
    var ans=[];
    for(s of smalls){
        if(dict.has(s)){
            ans.push(dict.get(s));
        }else{
            ans.push([]);
        }
    }
    return ans;
};

File: Python/面试题 17.17. 多次搜索.py
class Trie:
    def __init__(self,words):
        self.d={}
        for word in words:
            d=self.d
            for l in word:
                if l not in d:
                    d[l]={}
                d=d[l]
            d['end']=word
    
    def serach(self,word):
        
        d=self.d
        matches=[]
        for l in word:
            if l not in d:
                break
            d=d[l]
            if 'end' in d:
                matches.append(d['end'])
        return matches

class Solution:
    def multiSearch(self, big: str, smalls: List[str]) -> List[List[int]]:
        t=Trie(smalls)
        dict=defaultdict(list)


        for i in range(len(big)):
            matches=t.serach(big[i:])
            #print(matches)
            for match in matches:
                dict[match].append(i)
        
        ans=[]
        for word in smalls:
            ans.append(dict[word])
        return ans



